src/completion.ts

```diff
@@ -0,0 +1,118 @@
+import { Arg, Generator, Option, Spec, Subcommand } from "./types.ts";
+import { makeArray } from "./util.ts";
+
+export const JSON_START = "----- Begin JSON -----\n";
+export const JSON_END = "\n----- End JSON -----";
+
+export const completion: Subcommand = {
+  name: "completion",
+  description: "Generate Fig completions",
+  subcommands: [
+    {
+      name: "fig",
+      description: "Generates Fig completions",
+      async action() {
+      },
+    },
+    {
+      name: "run-generator",
+      description: "Runs a generator",
+      args: [
+        {
+          name: "lookup-path",
+          description: `The lookup path of the generator`,
+        },
+        {
+          name: "tokens",
+          description: "Base-64 encoding of a JSON array of strings",
+        },
+      ],
+      async action({ path, args: [lookupPathString, tokensBase64] }) {
+        const tokens = JSON.parse(atob(tokensBase64));
+        const generator = findGenerator(lookupPathString, path[0]);
+        if (!generator.custom) {
+          throw new Error("No `custom` property on generator");
+        }
+        const suggestions = await generator.custom(tokens);
+        const json = JSON.stringify(suggestions, null, 2);
+        console.log(`${JSON_START}${json}${JSON_END}`);
+      },
+    },
+  ],
+};
+
+export function findGenerator(lookupPathString: string, root: Spec): Generator {
+  const parts = lookupPathString.split(",");
+  let expect: "soa" | "a" | "g" = "soa";
```

Comment:
Add comment explaining the cleverness, the letters in this string are the valid letters. Will transition left to right through this union.
* First expect a string, option or arg.
* If you find an option, expect an arg.
* If you find an arg, expect a generator.